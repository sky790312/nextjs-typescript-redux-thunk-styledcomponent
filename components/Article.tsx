import * as React from 'react'
import { useDispatch } from 'react-redux'
import { ArticleSchema } from '@/schema/articleSchema'

type Props = {
  article: ArticleSchema
  removeArticle: (article: ArticleSchema) => void
}

export const Article: React.FC<Props> = ({ article, removeArticle }) => {
  const dispatch = useDispatch()

  const deleteArticle = React.useCallback(
    (currentArticle: ArticleSchema) => dispatch(removeArticle(currentArticle)),
    [dispatch, removeArticle]
  )

  return (
    <div className="Article">
      <div>
        <h1>{article.title}</h1>
        <p>{article.body}</p>
      </div>
      <button onClick={() => deleteArticle(article)}>Delete</button>
    </div>
  )
}
