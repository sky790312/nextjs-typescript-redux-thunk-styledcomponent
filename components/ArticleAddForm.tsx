import * as React from 'react'
import { ArticleSchema } from '@/schema/articleSchema'

type Props = {
  add: (article: ArticleSchema | any) => void
}

export const ArticleAddForm: React.FC<Props> = ({ add }) => {
  const [article, setArticle] = React.useState<
    ArticleSchema | Record<string, unknown>
  >()

  const handleArticleData = (e: React.FormEvent<HTMLInputElement>) => {
    setArticle({
      ...article,
      [e.currentTarget.id]: e.currentTarget.value,
    })
  }

  const addNewArticle = (e: React.FormEvent) => {
    e.preventDefault()
    add(article)
  }

  return (
    <form onSubmit={addNewArticle} className="Add-article">
      <input
        type="text"
        id="title"
        placeholder="Title"
        onChange={handleArticleData}
      />
      <input
        type="text"
        id="body"
        placeholder="Description"
        onChange={handleArticleData}
      />
      <button disabled={!article}>Add article</button>
    </form>
  )
}
