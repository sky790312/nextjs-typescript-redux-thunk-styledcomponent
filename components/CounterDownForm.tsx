import React, { useState } from 'react'
import { useInterval } from '@/hooks/useInterval'
import { DELAY_TIME } from '@/constants'
import { isNumberString } from '@/utils'
import styled from 'styled-components/'
import { useSelector, shallowEqual, useDispatch } from 'react-redux'
import { Dispatch } from 'redux'
import { counterActions } from '@/store/state.counter'
import { RootState } from '@/store/index'

export const CounterDownForm: React.FC = React.memo(() => {
  const dispatch: Dispatch = useDispatch()
  const { setTimer } = counterActions
  const [timerInputValue, setTimerInputValue] = useState<string>('')
  const timer = useSelector(
    (state: RootState) => state.counter.timer,
    shallowEqual
  )

  useInterval(() => {
    if (!timer) {
      return
    }

    dispatch(setTimer(timer - 1))
  }, DELAY_TIME)

  const onTimerInputChange = (e: React.FormEvent<HTMLInputElement>) => {
    const value = e.currentTarget.value

    if (value === '' || isNumberString(value)) {
      setTimerInputValue(value)
    }
  }

  const handleTimer = () => {
    if (!timerInputValue) {
      return
    }

    dispatch(setTimer(parseInt(timerInputValue, 10)))
    setTimerInputValue('')
  }

  return (
    <form
      onSubmit={(e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault()
        handleTimer()
      }}
    >
      <StyledInput
        type="text"
        placeholder="請輸入倒數時間(數字)"
        value={timerInputValue}
        onChange={onTimerInputChange}
      />
      分鐘
      <StyledButton onClick={handleTimer}>設定</StyledButton>
    </form>
  )
})

const StyledInput = styled.input`
  padding: 12px 40px;
`

const StyledButton = styled.button`
  padding: 12px 40px;
  margin-left: 20px;
  border-radius: 4px;
  border: none;
  cursor: pointer;
  color: ${({ theme }) => theme.colors.white};
  background-color: ${({ theme }) => theme.colors.gray};
`
