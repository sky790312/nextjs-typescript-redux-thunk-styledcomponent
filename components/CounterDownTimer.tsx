import React from 'react'
import { showYYMMSS } from '@/utils'
import styled from 'styled-components/'
import { useSelector, shallowEqual } from 'react-redux'
import { RootState } from '@/store/index'

export const CounterDownTimer: React.FC = () => {
  const timer = useSelector(
    (state: RootState) => state.counter.timer,
    shallowEqual
  )

  return <StyledCountDownTimer>{showYYMMSS(timer)}</StyledCountDownTimer>
}

const StyledCountDownTimer = styled.div`
  margin: 100px 0;
  font-size: 6rem;
  color: ${({ theme }) => theme.colors.blue};
`
