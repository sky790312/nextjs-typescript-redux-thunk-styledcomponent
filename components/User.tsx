import React, { useState, useCallback } from 'react'
import { UserSchema } from '@/schema/userSchema'
import styled from 'styled-components/'

type Props = {
  user: UserSchema
}

export const User: React.FC<Props> = ({ user }) => {
  return (
    <UserContainer>
      <Left>
        <UserImage name={user.name} />
      </Left>
      <Right>
        {user.id}. {user.name}
      </Right>
    </UserContainer>
  )
}

const UserContainer = styled.div`
  display: flex;
  align-items: center;
  margin: 20px;
  padding: 20px;
  border: 1px solid ${({ theme }) => theme.colors.black};
  background-color: ${({ theme }) => theme.colors.gray};
`

const Left = styled.div`
  padding: 20px;
`

const Right = styled.div`
  padding: 20px;
`

const UserImage = styled.div<{ name: string }>`
  width: 50px;

  &:before {
    background: ${({ theme }) => theme.colors.lightBlue};
    color: ${({ theme }) => theme.colors.white};
    opacity: 1;
    content: '${(props) => props.name}';
    display: inline-block;
    font-weight: 600;
    border-radius: 50%;
    vertical-align: middle;
    margin-right: 0.5em;
    width: 50px;
    height: 50px;
    line-height: 50px;
    text-align: center;
  }
`
