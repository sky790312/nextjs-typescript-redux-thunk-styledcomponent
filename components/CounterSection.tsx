import React from 'react'
import { CounterDownForm } from '@/components/CounterDownForm'
import { CounterDownTimer } from '@/components/CounterDownTimer'

export const CounterSection: React.FC = React.memo(() => {
  return (
    <div>
      <h1>抽獎時間</h1>
      <CounterDownForm />
      <CounterDownTimer />
    </div>
  )
})
