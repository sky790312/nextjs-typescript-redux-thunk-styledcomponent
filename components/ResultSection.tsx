import React from 'react'
import { useSelector, shallowEqual } from 'react-redux'
import styled from 'styled-components/'
import { RootState } from '@/store/index'
import { User } from '@/components/User'

export const ResultSection: React.FC = () => {
  const winnerUser = useSelector(
    (state: RootState) => state.users.winnerUser,
    shallowEqual
  )

  return (
    <div>
      <h1>抽獎結果</h1>
      {winnerUser && (
        <ResultContainer>
          <User user={winnerUser} />
          <h3>恭喜！！ 🎉🎉🎉</h3>
        </ResultContainer>
      )}
    </div>
  )
}
const ResultContainer = styled.div``
