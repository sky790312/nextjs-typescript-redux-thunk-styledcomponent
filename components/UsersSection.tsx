import React from 'react'
import { User } from '@/components/User'
import styled from 'styled-components/'
import { RootState } from '@/store/index'
import { useSelector, shallowEqual } from 'react-redux'

export const UsersSection: React.FC = () => {
  const users = useSelector(
    (state: RootState) => state.users.users,
    shallowEqual
  )

  return (
    <div>
      <h1>參與抽獎名單</h1>
      <UsersContainer>
        {users.map((user) => (
          <User key={user.id} user={user} />
        ))}
      </UsersContainer>
    </div>
  )
}
const UsersContainer = styled.div`
  max-height: 480px;
  overflow: auto;
`
