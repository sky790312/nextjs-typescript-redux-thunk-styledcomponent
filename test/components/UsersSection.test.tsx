import React from 'react'
import { render, withProviders } from '@/test/testUtils'
import { UsersSection } from '@/components/UsersSection'

describe('UsersSection component', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(withProviders(<UsersSection />), {})
    expect(asFragment()).toMatchSnapshot()
  })
})
