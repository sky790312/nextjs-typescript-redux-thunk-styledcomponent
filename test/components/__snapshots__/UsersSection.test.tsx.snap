// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersSection component matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1>
      參與抽獎名單
    </h1>
    <div
      class="sc-eCssSg hdMxnt"
    >
      <div
        class="sc-bdfBwQ fckRnc"
      >
        <div
          class="sc-gsTCUz dpEyqO"
        >
          <div
            class="sc-hKgILt eCVQLM"
            name="Kevin"
          />
        </div>
        <div
          class="sc-dlfnbm lgaWeB"
        >
          1. Kevin
        </div>
      </div>
      <div
        class="sc-bdfBwQ fckRnc"
      >
        <div
          class="sc-gsTCUz dpEyqO"
        >
          <div
            class="sc-hKgILt QiaJO"
            name="Steve"
          />
        </div>
        <div
          class="sc-dlfnbm lgaWeB"
        >
          2. Steve
        </div>
      </div>
      <div
        class="sc-bdfBwQ fckRnc"
      >
        <div
          class="sc-gsTCUz dpEyqO"
        >
          <div
            class="sc-hKgILt dXGOUY"
            name="Calvin"
          />
        </div>
        <div
          class="sc-dlfnbm lgaWeB"
        >
          3. Calvin
        </div>
      </div>
      <div
        class="sc-bdfBwQ fckRnc"
      >
        <div
          class="sc-gsTCUz dpEyqO"
        >
          <div
            class="sc-hKgILt klPyyw"
            name="Diana"
          />
        </div>
        <div
          class="sc-dlfnbm lgaWeB"
        >
          4. Diana
        </div>
      </div>
      <div
        class="sc-bdfBwQ fckRnc"
      >
        <div
          class="sc-gsTCUz dpEyqO"
        >
          <div
            class="sc-hKgILt elvIAn"
            name="Joe"
          />
        </div>
        <div
          class="sc-dlfnbm lgaWeB"
        >
          5. Joe
        </div>
      </div>
      <div
        class="sc-bdfBwQ fckRnc"
      >
        <div
          class="sc-gsTCUz dpEyqO"
        >
          <div
            class="sc-hKgILt bmcZWg"
            name="Jack"
          />
        </div>
        <div
          class="sc-dlfnbm lgaWeB"
        >
          6. Jack
        </div>
      </div>
      <div
        class="sc-bdfBwQ fckRnc"
      >
        <div
          class="sc-gsTCUz dpEyqO"
        >
          <div
            class="sc-hKgILt gTgzjB"
            name="Ray"
          />
        </div>
        <div
          class="sc-dlfnbm lgaWeB"
        >
          7. Ray
        </div>
      </div>
      <div
        class="sc-bdfBwQ fckRnc"
      >
        <div
          class="sc-gsTCUz dpEyqO"
        >
          <div
            class="sc-hKgILt dVoujR"
            name="Allen"
          />
        </div>
        <div
          class="sc-dlfnbm lgaWeB"
        >
          8. Allen
        </div>
      </div>
      <div
        class="sc-bdfBwQ fckRnc"
      >
        <div
          class="sc-gsTCUz dpEyqO"
        >
          <div
            class="sc-hKgILt mmURj"
            name="Zoe"
          />
        </div>
        <div
          class="sc-dlfnbm lgaWeB"
        >
          9. Zoe
        </div>
      </div>
      <div
        class="sc-bdfBwQ fckRnc"
      >
        <div
          class="sc-gsTCUz dpEyqO"
        >
          <div
            class="sc-hKgILt bYFysD"
            name="James"
          />
        </div>
        <div
          class="sc-dlfnbm lgaWeB"
        >
          10. James
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
